### PR TITLE
strbuf: Library that implements printing to a dynamically allocated buffer.

### PR DESCRIPTION
**Pull request by octo**
I'm planning on using this code to build the string representation of a label based metric. This certainly will have uses all throughout the daemon though. Looking for "ADD" or "ADDF" macros should uncover places that might benefit from this code.

ChangeLog: collectd: The `strbuf` utility library has been added. It simplifies building strings iteratively.

### CODE REVIEW

File: src/utils/strbuf/strbuf.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  &(strbuf_t) { .ptr = NULL }
+#define STRBUF_NEW_STATIC(b)                                                   \
+  &(strbuf_t) { .ptr = b, .size = sizeof(b), .fixed = 1 }
+#define STRBUF_DELETE(buf)                                                     \
+  do {                                                                         \
+    if (!buf->fixed) {                                                         \
+      free(buf->ptr);                                                          \
+    }                                                                          \
+    *buf = (strbuf_t){.ptr = NULL};                                            \
+  }
```
Comment (igorpeshansky) [inline src/utils/strbuf/strbuf.h:47]: These macros are public. Should you be testing them as well?

Comment (octo) [reply src/utils/strbuf/strbuf.h]: Excellent point, thank you! I've added tests and even found an error with the `STRBUF_DESTROY` macro.

**Comment by octo** [comment]
I added an update to `src/utils/common/common.c` which showcases how the `strbuf`library can be used to simplify existing code. While there, I also improved the code coverage of the "common" library.

File: src/utils/strbuf/strbuf.c [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    if (tmp >= 1)
+      cached_pagesize = (size_t)tmp;
+    else
+      cached_pagesize = 1024;
+  }
+
+  return cached_pagesize;
+}
+
+/* strbuf_resize resizes an dynamic buffer to ansure that "need" bytes can be
```
Comment (sunkuranganath) [inline src/utils/strbuf/strbuf.c:56]: 'to ensure'

Comment (octo) [reply src/utils/strbuf/strbuf.c]: Done, thank you!

**Comment by octo** [comment]
While using this library to simplify existing code, I noticed two more common use-cases:

1.  Limited print: provide a function that takes a size, printing only that number of characters. When extracting part of a string, printing everything up to the null byte forces users to copy the relevant part into a temporary buffer. The new `strbuf_printn` function addresses this use-case.
1.  Escaped print: provide a function that escapes the string before adding it to the buffer. When creating output strings, certain characters commonly need to be escaped, e.g. a `"` (quote) may need to be printed as `\"`. The new `strbuf_print_escaped` takes a list of characters that need to be escaped and an escape character, ensuring that each to-be-escaped character is preceded by the escape character.

    I'm a bit conflicted whether letting the caller provide a list of "forbidden" characters and an escape character is the best approach here. The existing `escape_string` function always uses `\` (backslash) to escape and has a fixed set of characters it escapes, including `\n` (newline) and `"` (quote). I.e. the flexibility may not be needed. On the other hand, code such as `format_json` have very specific escaping rules and implement their own functions for this. What do you think?

Thanks and best regards,
—octo

**Comment by mrunge** [comment]
Looks good to me!

**Comment by octo** [comment]
Thanks Matthias!

**Pull request merged**